Extract: Improved region-outlining for complex procedures

mlange05 opened the pull request:
This change brings various improvements to enable the use of `outine_pragma_regions` for more complex code regions. This is intended to eventually allow us to combine pragma-based extraction with auto-parallelisation to re-rewrite complex control-flow sections.

In particular, this PR refactors the utility structure to allow invocations per-region, deals with derived type symbols and arguments, and imposes an ordering on extracted argument signatures. There are also other small improvements added, as well as a utility to remove unused imports and import symbols.

In more detail:
* Use of `Transformer`, instead of the more complex `NestedMaskedTransformer` to replace `PragmaRegion` nodes
* Separation of the outer loop over subroutines from the specific single call-subroutine extraction for a single region
* A new utility to remove unused imports and trim unused symbols from import statements within a routine
* Ability to handle C-style imports when pragma-outlining sburoutines
* Resolving explicitly requested argument symbols (via the "intent" notation) from the parent routine. This allows explicitly requesting arguments that might not be used in the body, but are necessary for shape info in the spec of the extracted routine.
* Handling of derived-type symbols in extracted body/routine, in particular ensuring that only the parent is added to the extracted routine signature
* Additional tests for derived-type arguments and associated symbols in extracted bodies/routines. The latter is not strictly needed, but useful to have.
* Explicit ordering of arguments for reproducibility - this will use lexicographic base ordering, but put derived-type arguments, before array arguments, before scalar arguments.

reuterbal commented on loki/transformations/remove_code.py:
```diff
@@ -332,3 +337,68 @@ def visit_Import(self, o, **kwargs):
         return self._rebuild(o, rebuilt)
+
+
+def do_remove_unused_imports(routine):
```
Is this not redundant to or at least overlapping in functionality with the `sanitise_imports` utility? I would prefer if we combined the two somehow.
https://github.com/ecmwf-ifs/loki/blob/83c9ebe002064e0544a4b3bf0261ca1bbe9323ee/loki/transformations/utilities.py#L281-L295

reuterbal reviewed (changes requested):
This looks great and is a valuable expansion of functionality over the original stubs!

I've left a few minor comments that would be good to address but no major showstoppers. However, I think we have a duplication between `sanitise_imports` and the new utility to remove unused imports.

mlange05 commented on loki/transformations/extract/outline.py:
```diff
@@ -6,18 +6,155 @@
+__all__ = ['outline_region', 'outline_pragma_regions']
+
+
+def order_variables_by_type(variables, imports=None):
```
😏

mlange05 commented:
@reuterbal Ok, I've rebased out the import sanitisation utility for now and addressed all other comments in the top two commits, as we as a fresh rebase over main. Please have another look.

reuterbal reviewed (approved):
Excellent, looks good.

I agree on the better implementation of the import sanitisation and happy to sort out the overlap separately!
